File: MapReduce_FastQ_BaseQual_Baseline/src/main/java/baseQual/FastQ_PerBaseQual_Job.java
package baseQual;

import org.apache.hadoop.conf.Configuration;
import org.apache.hadoop.fs.Path;
import org.apache.hadoop.io.DoubleWritable;
import org.apache.hadoop.io.IntWritable;
import org.apache.hadoop.mapreduce.Job;
import org.apache.hadoop.mapreduce.lib.input.FileInputFormat;
import org.apache.hadoop.mapreduce.lib.output.FileOutputFormat;
import org.seqdoop.hadoop_bam.FastqInputFormat;
import utils.IdentifierPositionKeyWritable;

/**
 * master-thesis Clemens Banas
 * Organization: DBIS - University of Innsbruck
 * Created 04.02.16.
 */
public class FastQ_PerBaseQual_Job {

    public static void main(String[] args) throws Exception {
        Configuration conf = new Configuration();
        conf.set("mapreduce.reduce.java.opts", "-Xmx6500M");
        Job job = Job.getInstance(conf, "MR_FastQ_PerBaseQual_Baseline");

        job.setJarByClass(FastQ_PerBaseQual_Job.class);
        job.setMapperClass(FastQ_PerBaseQual_Mapper.class);
        job.setReducerClass(FastQ_PerBaseQual_Reducer.class);

        job.setInputFormatClass(FastqInputFormat.class);
        job.setMapOutputKeyClass(IdentifierPositionKeyWritable.class);
        job.setMapOutputValueClass(IntWritable.class);

        job.setOutputKeyClass(IdentifierPositionKeyWritable.class);
        job.setOutputValueClass(DoubleWritable.class);

        job.setOutputFormatClass(FastQ_PerBaseQual_OutputFormat.class);

        FileInputFormat.addInputPath(job, new Path(args[0]));
        FileOutputFormat.setOutputPath(job, new Path(args[1]));

        if (!job.waitForCompletion(true)) {
            System.err.println("sort :: Job failed.");
        }
    }

}

File: Spark_FastQ_BaseQual_Baseline/src/main/scala/FastQ_PerBaseQual_Job.scala
package main.scala

import main.scala.utils.AvgCount
import org.apache.hadoop.conf.Configuration
import org.apache.hadoop.io.Text
import org.apache.hadoop.mapreduce.lib.input.FileSplit
import org.apache.spark.rdd.{NewHadoopRDD, RDD}
import org.apache.spark.{SparkConf, SparkContext}
import org.seqdoop.hadoop_bam.{FastqInputFormat, SequencedFragment}

/**
  * master-thesis Clemens Banas
  * Organization: DBIS - University of Innsbruck
  * Created 21.04.2016
  */
object FastQ_PerBaseQual_Job {

  def main(args: Array[String]): Unit = {
    if (args.length != 2) {
      println("usage: spark-submit Spark_FastQ_BaseQual_Baseline-1.0-SNAPSHOT.jar <fastQ input file> <output dir>")
      return;
    }

    val input = args(0)
    val output = args(1)

    val conf = new SparkConf()
    conf.setAppName("Spark_FastQ_PerBaseQual_Baseline")
    conf.set("spark.serializer", "org.apache.spark.serializer.KryoSerializer")
    conf.registerKryoClasses(Array(classOf[AvgCount]))
    val sc = new SparkContext(conf)

    val configuration = new Configuration()
    configuration.set("mapreduce.input.fileinputformat.inputdir", input)

    val fastQFileRDD = sc.newAPIHadoopRDD( //reads a fastQ file from HDFS
      configuration,
      classOf[FastqInputFormat],
      classOf[Text],
      classOf[SequencedFragment]
    )

    //retrieve the filename
    val hadoopRdd = fastQFileRDD.asInstanceOf[NewHadoopRDD[Text, SequencedFragment]]
    val myRdd: RDD[Pair[String, SequencedFragment]] = hadoopRdd.mapPartitionsWithInputSplit { (inputSplit, iterator) ⇒
      val file = inputSplit.asInstanceOf[FileSplit]
      iterator.map { record ⇒ (file.getPath.getName, record._2) }
    }

    //mapping step
    val qualityScores: RDD[Pair[Pair[String, Int], Int]] = myRdd.flatMap( record => FastQ_PerBaseQual_Mapper.flatMap(record._1, record._2) )

    //reduce step
    val countedQualityScores: RDD[Pair[Pair[String, Int], AvgCount]] = qualityScores.combineByKey(
      (qualVal: Int) => FastQ_PerBaseQual_Reducer.createAverageCount(qualVal),
      (a: AvgCount, qualVal: Int) => FastQ_PerBaseQual_Reducer.addAndCount(a, qualVal),
      (a: AvgCount, b: AvgCount) => FastQ_PerBaseQual_Reducer.combine(a, b)
    )

    val res: RDD[Pair[Pair[String, Int], Double]] = countedQualityScores.map( record => FastQ_PerBaseQual_Mapper.map(record) );

    res.map( record => record._1._1 + "," + record._1._2 + "," + record._2 ).saveAsTextFile(output)
  }

}